server/src/services/backup.service.ts

```diff
@@ -0,0 +1,89 @@
+import { Injectable } from '@nestjs/common';
+import { exec } from 'node:child_process';
+import { OnEvent } from 'src/decorators';
+import { DatabaseLock } from 'src/interfaces/database.interface';
+import { ArgOf } from 'src/interfaces/event.interface';
+import { JobName, JobStatus } from 'src/interfaces/job.interface';
+import { BaseService } from 'src/services/base.service';
+import { handlePromiseError } from 'src/utils/misc';
+import { validateCronExpression } from 'src/validation';
+
+@Injectable()
+export class BackupService extends BaseService {
+  private backupLock = false;
+
+  @OnEvent({ name: 'app.bootstrap' })
+  async onBootstrap() {
+    const config = await this.getConfig({ withCache: false });
+
+    const { database } = config.backups;
+
+    // This ensures that backups only occur in one microservice
+    this.backupLock = await this.databaseRepository.tryLock(DatabaseLock.BackupDatabase);
+
+    if (this.backupLock) {
+      this.jobRepository.addCronJob(
```

Comment:
What does this mean in a scenario where there's multiple microservices workers (eg a scaled setup) and the one that got this lock is removed? Is the cron job in memory meaning it will no longer run? And are there any other (cron)jobs that applies to?

Reply:
This is a fair point, but an issue we already have with the setup. I am not sure how we would easily resolve this right now. I'm not sure if we explicitly support dynamic scaling right now. Maybe we should force cron creation to run on the api server, as they are less likely to be dynamically scaled I would expect.